Terraria/Utilities/SlotVector.cs

```diff
@@ -0,0 +1,195 @@
+		}
+		public bool Has(int index)
+		{
+			return index >= 0 && index < this._array.Length && this._array[index].Id.IsActive;
+		}
+		public bool Remove(SlotId id)
+		{
+			if (id.IsActive)
+			{
+				uint index = id.Index;
+				this._array[(int)((UIntPtr)index)] = new SlotVector<T>.ItemPair(default(T), id.ToInactive(this._freeHead));
+				this._freeHead = index;
+				this._count--;
+				return true;
+			}
+			return false;
+		}
+		IEnumerator<SlotVector<T>.ItemPair> IEnumerable<SlotVector<T>.ItemPair>.GetEnumerator()
+		{
+			return new SlotVector<T>.Enumerator(this);
+		}
+		IEnumerator IEnumerable.GetEnumerator()
+		{
+			return new SlotVector<T>.Enumerator(this);
+		}
+		public int Capacity
+		{
			get
+			{
+				return this._array.Length;
+			}
+		}
+		public int Count
+		{
			get
+			{
+				return this._count;
+			}
+		}
+		public T this[int index]
+		{
			get
+			{
+				if (index < 0 || index >= this._array.Length)
+				{
+					throw new IndexOutOfRangeException();
+				}
+				if (!this._array[index].Id.IsActive)
+				{
+					throw new KeyNotFoundException();
+				}
+				return this._array[index].Value;
+			}
			set
+			{
+				if (index < 0 || index >= this._array.Length)
+				{
+					throw new IndexOutOfRangeException();
+				}
+				if (!this._array[index].Id.IsActive)
+				{
+					throw new KeyNotFoundException();
+				}
+				this._array[index] = new SlotVector<T>.ItemPair(value, this._array[index].Id);
+			}
+		}
+		public T this[SlotId id]
+		{
			get
+			{
+				uint index = id.Index;
+				if (index < 0u || (ulong)index >= (ulong)((long)this._array.Length))
+				{
+					throw new IndexOutOfRangeException();
+				}
+				if (!this._array[(int)((UIntPtr)index)].Id.IsActive || id != this._array[(int)((UIntPtr)index)].Id)
+				{
+					throw new KeyNotFoundException();
+				}
+				return this._array[(int)((UIntPtr)index)].Value;
+			}
			set
+			{
+				uint index = id.Index;
+				if (index < 0u || (ulong)index >= (ulong)((long)this._array.Length))
+				{
+					throw new IndexOutOfRangeException();
+				}
+				if (!this._array[(int)((UIntPtr)index)].Id.IsActive || id != this._array[(int)((UIntPtr)index)].Id)
+				{
+					throw new KeyNotFoundException();
+				}
+				this._array[(int)((UIntPtr)index)] = new SlotVector<T>.ItemPair(value, id);
+			}
+		}
+		private const uint MAX_INDEX = 65535u;
+		private SlotVector<T>.ItemPair[] _array;
+		private int _count;
+		private uint _freeHead;
+		private uint _usedSpaceLength;
+		public class Enumerator : IEnumerator<SlotVector<T>.ItemPair>, IDisposable, IEnumerator
+		{
			public Enumerator(SlotVector<T> slotVector)
+			{
+				this._slotVector = slotVector;
+			}
+			public bool MoveNext()
+			{
+				while ((long)(++this._index) < (long)((ulong)this._slotVector._usedSpaceLength))
+				{
+					if (this._slotVector.Has(this._index))
+					{
+						return true;
+					}
+				}
+				return false;
+			}
+			public void Reset()
+			{
+				this._index = -1;
```

Comment:
All these casts to long and ulong seem like they're unnecessary, maybe investigate what's going on there